### PR TITLE
Enhance InPlacePodVerticalScaling performance

### PR DESCRIPTION
**Pull request by horacexd**
What type of PR is this?
/kind enhancement
/kind bug

What this PR does / why we need it:
It enhance the performance on InPlaceUpdateVerticalScaling
Below shows the doc link, and this PR is to fix the 3rd item
https://docs.google.com/document/d/1V3DLh3pH3CD-xhhJvAnOq_oWgPyjO-vj6wY6qdew9H0/edit#heading=h.5gfeh6i9awgb

Which issue(s) this PR fixes:
Fixes #

Special notes for your reviewer:
Does this PR introduce a user-facing change?
NONE
Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
enhancement report: https://docs.google.com/document/d/1Jxgezb5pcstB_UuTlpJWa8zmgwABK4fssiV-l7ljLeQ/edit?usp=sharing

### CODE REVIEW

**Comment by Jeffwan** [comment]
/hold 

Let's hold this change. The upstream is slightly different from our downstream. We need more testing on this PR.

**Comment by pacoxu** [comment]
Are there any benchmarks to prove the performance improvements? At least, some test data will help the end user understand.

**Comment by Jeffwan** [comment]
@pacoxu Sure. We will add some before and after logs and metrics we have in the issue.

**Comment by Jeffwan** [comment]
We rebased the master and did one more round testing yesterday and the performance meets the expectation, we can unhold this story. This PR address issue https://github.com/kubernetes/kubernetes/issues/112264. 

/hold cancel

**Review by Jeffwan — APPROVED** [review]
The change looks good to me. We are running a downstream version with this change and it helps a lot improving the performance. Please help review it.

/cc @pacoxu @SergeyKanzhelev @bobbypage @dchen1107 @mrunalp

**Comment by Jeffwan** [comment]
@vinaykul @mrunalp Do you get chance to look at the improvement?

**Comment by vinaykul** [comment]
@horacexd @Jeffwan Thanks for identifying this issue and the doc that clearly calls out the issue. I'll make some time to take a close look at the fix.

But on first glance (I might have missed it), I don't think there is an issue tracking this bug with extended resources. Can you please create the issue and link it in the description above? This will also need a E2E test case ( can be added to test/e2e/node/pod_resize.go )
thanks again.

**Comment by Jeffwan** [comment]
@vinaykul It addresses issues in https://github.com/kubernetes/kubernetes/issues/112264. I didn't create new issues

**Comment by mrunalp** [comment]
@bobbypage ptal

**Comment by vinaykul** [comment]
> #112264

Ah. The doc link in the description takes me to the 4th issue (custom resources) and that threw me off. @horacexd Please fix the link.

**Comment by Jeffwan** [comment]
@SergeyKanzhelev @mrunalp please help add 1.29 milestone label and thank you!

**Comment by horacexd** [comment]
> > #112264
> 
> Ah. The doc link in the description takes me to the 4th issue (custom resources) and that threw me off. @horacexd Please fix the link.

sry for the miss leading. The link is updated

**Comment by Jeffwan** [comment]
@vinaykul @pacoxu  @MaryamTavakkoli  this is kind of critical on the performance side,  can we include this one in v1.29?

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1921,7 +1921,7 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 		// TODO(vinaykul,InPlacePodVerticalScaling): Investigate doing this in HandlePodUpdates + periodic SyncLoop scan
 		//     See: https://github.com/kubernetes/kubernetes/pull/102884#discussion_r663160060
 		if kl.podWorkers.CouldHaveRunningContainers(pod.UID) && !kubetypes.IsStaticPod(pod) {
-			pod = kl.handlePodResourcesResize(pod)
+			pod = kl.handlePodResourcesResize(pod, &apiPodStatus)
```
Comment (liggitt) [inline pkg/kubelet/kubelet.go:1924]: cc @smarterclayton @bobbypage for data flow of API pod status and kubelet-constructed pod status ... this looks like it might cause more issues like the ones identified in https://github.com/kubernetes/kubernetes/issues/116971

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1947,6 +1947,14 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 	}
 
 	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && isPodResizeInProgress(pod, &apiPodStatus) {
+		// Before syncPod end, get the latest podStatus from CRI and refresh the pod
```
Comment (smarterclayton) [inline pkg/kubelet/kubelet.go:1950]: Why is this necessary?  This is normally the responsibility of the pod handler, and refreshing status generally needs a reason.

**Comment by MaryamTavakkoli** [comment]
> @vinaykul @pacoxu @MaryamTavakkoli this is kind of critical on the performance side, can we include this one in v1.29?

I will seek advice from @hailkomputer on this.

**Comment by hailkomputer** [comment]
> @vinaykul @pacoxu @MaryamTavakkoli this is kind of critical on the performance side, can we include this one in v1.29?

Hello, as we are in currently in the code freeze for v1.29 and as it was remarked in the review comments by @liggitt that the data flow might be introducing issues, we strongly believe that it is better to leave the PR out of the milestone.

**Comment by vinaykul** [comment]
@Jeffwan Let's target to do this one in earlier part of next release. Please also have a look at @smarterclayton document when you get the chance.
https://docs.google.com/presentation/d/1wbPFANXVbk-e-5BvVHaZScqoqJsfcU_FQVmG-N7pv3c/

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (LastNight1997) [reply pkg/kubelet/kubelet.go:1924]: We have three scenarios where we need to make this modification：

Case 1:
1. Enter SyncPod for the first time after modifying Pod Resource. Inplace VPA is successful, but update podStatus is asynchronous, and the resize/resource status in APIServer has not been successfully updated.
2. At this time, if the second syncPod is triggered immediately(for some other reason, eg. pod update during first SyncPod), the resize status obtained from APIServer is still InProgess and the resource in podStatus is still old. At this time, the status is updated again by handlePodResourcesResize to InProgess and the InPlaceVPA process is repeated, but in the end when updating PodStatus, it was intercepted by isPodResizeInProgress(resource in spec is equal to container runtime) and could not complete the update.
3. One minute later, when syncPod comes in for the third time, the Resource in podStatus is already the latest, and the update is completed normally.

Solution: We fix this bug by comparing Spec Resource and Container runtime resource in handlePodResourcesResize (as performed in isPodResizeInProgress) to determine whether Resize is needed.

Case 2: 
Problem: When a pod enters the deferred state, if it is updated again, and the pod with new resource should still be in the deferred state. At this case, it will cause that the resize state of the pod will remain in Proposed instead of deferred.

Reason: After the Pod in the deferred status is update again, the apiserver will modify the status.resize to proposed, but the resize status in kubelet statusManager is still deferred. Then when kl.statusManager.SetPodStatus is call locally again to update resize status to deferred, it is found that the status is equal to cache in statusManager and will not be updated.

Solution: Modify determinePodResizeStatus, based on the status in apiserver, so that statusManager can update the status to proposed normally, so there will be no situation where it is inconsistent with the cache in statusManager and cannot be updated.

Case 3: 
Problem: If you only modify the Mem limit to below the current usage, the Pod resize status will also be stuck in the Proposed stage instead of InProgress.

Reason: handlePodResourcesResize determines whether to modify the status to InProgress only based on whether Request is equal to ResourcesAllocated.

Solution: Modify determinePodResizeStatus, as long as pod resize status is Proposed, allow it be changed to InProgress.

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (LastNight1997) [reply pkg/kubelet/kubelet.go:1950]: After calling CRI to update the Resource, there is no chance to update the Pod's ResizeStatus to completed. It need to wait until the next SyncPod, which is usually 1 minute later.

**Comment by Jeffwan** [comment]
@horacexd did you get a chance to take a look at the slide? Can you summarize the key state changes introduced in the PR? Based on my understanding, it just patches the status and that won't add/change the state machine.  Please double check whether it affects the next reconcile loop

**Comment by horacexd** [comment]
> @horacexd did you get a chance to take a look at the slide? Can you summarize the key state changes introduced in the PR? Based on my understanding, it just patches the status and that won't add/change the state machine. Please double check whether it affects the next reconcile loop

Yes, I've read the slides several times. Let me summarize the change so we could have further discussion.

**Comment by horacexd** [comment]
The change we made will let SyncPod to directly request CRI and update podStatus 

On [slide 7](https://docs.google.com/presentation/d/1wbPFANXVbk-e-5BvVHaZScqoqJsfcU_FQVmG-N7pv3c/edit#slide=id.g1de8a1ca1a4_0_1505), it recommended to use podWorker for sync action. Since podworker is runtime, I think it is OK to request status from CRI which satisfy this rule. 

After that, we convert status to API pod status and update statusManager, this also doesn't break the state machine on [slide 3](https://docs.google.com/presentation/d/1wbPFANXVbk-e-5BvVHaZScqoqJsfcU_FQVmG-N7pv3c/edit#slide=id.g1de8a1ca1a4_0_1377) 

For admission on in-place update, the biggest concern is that deferred state is more like a medium state, and it doesn't mark as "failed" on admission. I think this part is more like a desgin issue since the desired pod status passed admission but actual pod status cannot reach it.

Overall, I think this change doesn't break the state machine. And it should not block alpha release

Please correct me if I make any mistake or miss any thing.

cc: @Jeffwan @smarterclayton @liggitt

**Comment by smarterclayton** [comment]
/hold

I'd like to deeply review this, and I am currently not able to.  SyncPod is not expected to call CRI for dynamic status - that breaks a lot of status assumptions.  Will try to get this back into the review queue before end of next week.

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/kubelet/kubelet.go:1969]: That sounds incorrect.  If we make a CRI call, that updates the state of the object, then we should get a CRI event afterwards indicating the container runtime status changed (something should be different).  That should then trigger a subsequent SyncPod, which gives you a chance to react.  Thats <1-2s (and will be faster when the continuous pod status loop is added).

If you are seeing 1m waits (meaning you had to wait for house cleaning and sync), that means the change to the CRI either didn't generate an event, or *should* have generated an event and we didn't see it.  Can you verify for me whether the CRI update to the container actually changes the observable state of the container in such a way that it shows up in in the pleg?  Without this code, when you run SyncPod, I expect to see PLEG detect a container change.

If that's the problem, and we fix that, and it only takes 1-2s, this is no different from any other kubelet code, and we wouldn't need this change?

Comment (Jeffwan) [reply pkg/kubelet/kubelet.go:1993]: @smarterclayton 

1. Agree that `updateStatus` again in the same `kubelet.SyncPod` loop in not the  right way to go. 
2. PLEG currently can not differentiate the pod just finished resizing because container state is always "running" and the hash doesn't change (hash indicate the). There's no way to generate a `ContainerChanged` pleg event. There's definitely bug in the original VPA implementation. 

I am trying to fix PLEG logic for in-place VPA now.

**Review by Jeffwan — CHANGES_REQUESTED** [review]
